backends/bmv2/jsonconverter.cpp

```diff
@@ -101,6 +103,90 @@ class ErrorCodesVisitor : public Inspector {
+    JsonConverter* converter;
+    using Input = std::vector<const IR::Expression *>;
+    std::map<const IR::Declaration_Instance *, Input> selector_input_map{};
+
+ public:
+    explicit SharedActionSelectorCheck(JsonConverter* converter) : converter(converter)
+    { CHECK_NULL(converter); }
+
+    const Input &get_selector_input(const IR::Declaration_Instance* selector) const {
+      return selector_input_map.at(selector);
+    }
+
+    bool preorder(const IR::P4Table* table) override {
+        auto v1model = converter->v1model;
+        auto refMap = converter->refMap;
+        auto typeMap = converter->typeMap;
+
+        auto implementation = table->properties->getProperty(
+            v1model.tableAttributes.tableImplementation.name);
+        if (implementation == nullptr) return false;
+        if (!implementation->value->is<IR::ExpressionValue>()) {
+          ::error("%1%: expected expression for property", implementation);
+          return false;
+        }
+        auto propv = implementation->value->to<IR::ExpressionValue>();
+        if (!propv->expression->is<IR::PathExpression>()) return false;
+        auto pathe = propv->expression->to<IR::PathExpression>();
+        auto decl = refMap->getDeclaration(pathe->path, true);
+        if (!decl->is<IR::Declaration_Instance>()) {
+            ::error("%1%: expected a reference to an instance", pathe);
+            return false;
+        }
+        const auto &apname = decl->externalName();
+        auto dcltype = typeMap->getType(pathe, true);
+        if (!dcltype->is<IR::Type_Extern>()) {
+            ::error("%1%: unexpected type for implementation", dcltype);
+            return false;
+        }
+        auto type_extern_name = dcltype->to<IR::Type_Extern>()->name;
+        if (type_extern_name != v1model.action_selector.name) return false;
+
+        auto key = table->getKey();
+        Input input;
+        for (auto ke : *key->keyElements) {
+            auto mt = refMap->getDeclaration(ke->matchType->path, true)->to<IR::Declaration_ID>();
+            BUG_CHECK(mt != nullptr, "%1%: could not find declaration", ke->matchType);
+            if (mt->name.name != v1model.selectorMatchType.name) continue;
+            input.push_back(ke->expression);
+        }
+        auto decl_instance = decl->to<IR::Declaration_Instance>();
+        auto it = selector_input_map.find(decl_instance);
+        if (it == selector_input_map.end()) {
+            selector_input_map[decl_instance] = input;
+            return false;
+        }
+
+        // returns true if inputs are the same, false otherwise
+        auto cmp_inputs = [](const Input &i1, const Input &i2) {
+            for (auto e1 : i1) {
+                auto cmp_e = [e1](const IR::Expression *e2) {
+                    // not the best solution but the best available one given that there is no "depp
+                    // comparison" of expressions for now.
+                    return e1->toString() == e2->toString();
```

Comment:
The desired behavior is for the expressions to be the same, not the result of the evaluation. Which is why ideally the selector input would be a property of the action_selector, not the table. Having the selector input has part of the match key, even with a special match type and even if it does contribute to choosing an action entry, seems "wrong", as logically it is really a 2-step selection process: first you choose a group using a regular match, then you choose a member in the group by computing a hash over the selector input.
I tried coming up with an alternative extern definition for action_selector, but couldn't come up with anything within the boundaries of the current syntax (basically externs cannot "store" references to other P objects like fields), so this work-around works for me in the meantime.